test/integration/apiserver/watchcache_test.go

```diff
@@ -234,3 +238,64 @@ func BenchmarkListFromWatchCache(b *testing.B) {
 		b.Logf("Number of secrets: %d", len(secrets.Items))
 	}
 }
+
+func TestConsistentList(t *testing.T) {
+	var testcases = []struct {
+		name      string
+		fromCache bool
+	}{
+		{
+			name:      "FromCache",
+			fromCache: true,
+		},
+		{
+			name:      "FromStorage",
+			fromCache: false,
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			_, ctx := ktesting.NewTestContext(t)
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ConsistentListFromCache, tc.fromCache)()
+
+			clientSet, _, tearDownFn := framework.StartTestServer(ctx, t, framework.TestServerSetup{})
+			defer tearDownFn()
+
+			ns := framework.CreateNamespaceOrDie(clientSet, "list", t)
+			defer framework.DeleteNamespaceOrDie(clientSet, ns, t)
+
+			for i := 0; i < 10; i++ {
+				rs := newRS(ns.Name)
+				start := time.Now()
+				newRs, err := clientSet.AppsV1().ReplicaSets(ns.Name).Create(ctx, rs, metav1.CreateOptions{})
+				if err != nil {
+					t.Fatal(err)
+				}
+				deployments, err := clientSet.AppsV1().Deployments(ns.Name).List(ctx, metav1.ListOptions{})
+				duration := time.Since(start)
+				if err != nil {
+					t.Fatal(err)
+				}
+				deploymentsVersion, err := strconv.Atoi(deployments.ResourceVersion)
+				if err != nil {
+					t.Fatal(err)
+				}
+				replicaSetVersion, err := strconv.Atoi(newRs.ResourceVersion)
+				if err != nil {
+					t.Fatal(err)
+				}
+				if deploymentsVersion < replicaSetVersion {
+					t.Errorf("Expected list to have revision not lower then last write, deplyments: %q, replicaSet: %q", deploymentsVersion, replicaSetVersion)
+				}
+				if duration > time.Second/2 {
```

Comment:
I'm worried about this check - it may not be satisfied in overloaded environemnts, which our CI is...